core/src/main/java/io/substrait/expression/doc-files/FunctionArg.md

```diff
@@ -0,0 +1,54 @@
+## Representing and mapping Function arguments
+- `FunctionArg` is a marker interface
+- subtypes are `Expression`, `Type`, and  the new `EnumArg` class
+
+### EnumArg class
+- captured as the `Option` value and a reference to the `Enum`
+- Reference to the `Enum` is a reference to an `SimpleExtension.EnumArgument` instance.
+  - TODO: should also include the `FunctionAnchor`
+
+## Changes to Function invocation structure
+- `ScalarFunctionInvocation`, `AggregateFunctionInvocation`
+  - `arguments` is now  of type `List<FunctionArg>`
+  - currently retained `List<Expression> exprArguments()` which is a view on `arguments`
+    - seems useful. Need feedback..
+    - TODO: memonize
+
+## Proto mapping
+- **To Proto**
+  - map substrait `Expression` and `Type` instances to proto and then wrap them in a `FunctionArgument`
+  - map an `EnumArg` instance to a `FunctionArgument.Enum` that only holds the option value.
+- **From Proto**
+  - map `FunctionArgument`s wrapping `Type` and `Expression` to substrait `Type and `Expression` respectively
+  - for an `FunctionArgument.Enum` value at a particular parameter index in the function invocation
+    - look within possible `SimpleExtension.EnumArgument` and pick the first one that has the option value
```

Comment:
Does an Enum work similar to a Function in that its definition (i.e. name and possible values) is externalized and referenced by position through an anchor? I'd expect the String literal value to be captured plus a reference by index to the Enum definition. Something like that. I don't quite understand the necessity of all the search code you've written.

Reply:
In `algebra.proto` `FunctionArgument.Enum` only captures the enum value.
I took this as a design constraint.
So mapping when mapping to substrait have to also account for optional arguments.
Hence the search logic.

Reply:
Don't we have the type of the Enum also (based on the type of the expression)?

Reply:
This is the comment in `algebra.proto`
```
// deprecated: enum literals are only sensible in the context of
    // function arguments, for which FunctionArgument should now be
    // used
    Enum enum = 10 [deprecated = true];
```